MongoDB.FSharp/Serializers.fs

```diff
@@ -25,64 +24,67 @@ module Serializers =
-            let itemOptions = serializationOptions.ItemSerializationOptions
             let readArray() =
                 seq {
-                    reader.ReadStartArray()
+                    context.Reader.ReadStartArray()
                     let convention = BsonSerializer.LookupDiscriminatorConvention(typeof<'T>)
-                    while reader.ReadBsonType() <> BsonType.EndOfDocument do
-                        let actualElementType = convention.GetActualType(reader, typeof<'T>)
+                    while context.Reader.ReadBsonType() <> BsonType.EndOfDocument do
+                        let actualElementType = convention.GetActualType(context.Reader, typeof<'T>)
                         let serializer = BsonSerializer.LookupSerializer(actualElementType)
-                        let element = serializer.Deserialize(reader, typeof<'T>, itemOptions)
+                        let element = serializer.Deserialize(context, args)
                         yield element :?> 'T
-                    reader.ReadEndArray()
+                    context.Reader.ReadEndArray()
                 }
 
             let readArrayFromObject () =
-                reader.ReadStartDocument()
-                reader.ReadString("_t") |> ignore
-                reader.ReadName("_v")
-                let value = this.Deserialize(reader, actualType, actualType, options)
-                reader.ReadEndDocument()
+                context.Reader.ReadStartDocument()
+                context.Reader.ReadString("_t") |> ignore
+                context.Reader.ReadName("_v")
+                let value = this.Deserialize(context, args)
+                context.Reader.ReadEndDocument()
                 value
             
-            let bsonType = reader.GetCurrentBsonType()
+            let bsonType = context.Reader.GetCurrentBsonType()
             match bsonType with
             | BsonType.Null ->
-                reader.ReadNull()
-                null
-            | BsonType.Array -> readArray() |> List.ofSeq :> Object
+                context.Reader.ReadNull()
+                Unchecked.defaultof<'T list>
+            | BsonType.Array -> (readArray() |> List.ofSeq :> obj) :?> 'T list
             | BsonType.Document -> readArrayFromObject ()
             | _ -> 
-                let msg = sprintf "Can't deserialize a %s from BsonType %s" actualType.FullName (bsonType.ToString())
+                let msg = sprintf "Can't deserialize a %s from BsonType %s" args.NominalType.FullName (bsonType.ToString())
                 raise(InvalidOperationException(msg))
+        
+        override this.Serialize(context:BsonSerializationContext,  args:BsonSerializationArgs, value:'T list) =
+            if value = Unchecked.defaultof<'T list> then
+                // There aren't supposed to be null values in F#
+                context.Writer.WriteStartArray()
+                context.Writer.WriteEndArray()
+            else
+                let actualType = value.GetType()
```

Comment:
Value isn't used. I'm guessing that you still need the VerifyTypes or some equivalent method call